avramd - add_bootp contrib

avramd opened the pull request:
This PR is to capture modifications of @colin-foster-in-advantage's `add_bootp` proposal coming from discussion on his [systemd BOOTP support RFE PR](https://github.com/systemd/systemd/pull/34888).

Colin's branch is currently based on a commit in main in which `mkosi` is broken. I've rebased that branch onto a fairly recent main, in a new branch called `add_bootp-rebased` so that this PR can show only my diffs with Colin's work. 

Objectives:

- fix integration test broken in Colin's branch
- Address errors related to ignoring BOOTP `siaddr`
  - Update the implementation so that bootp unit tests will correctly fail when `server_address` is missing
  - Add a bootp integration test; this should also fail when `server_address` is missing
  - Update the bootp implementation to use the BOOTP `siaddr` field as `server_address` when `BOOTP=yes`

colin-foster-in-advantage commented:
Oh you got an integration test done! Awesome! I didn't think that was possible.

I'm new to `systemd` PRs, so I didn't know if rebasing was going to make things more difficult. I did it once though, and nobody complained.

I'll run a verification hopefully in the next 24 hours and merge things in. When I tried the patch you suggested last week I was getting crashes... then I lost some time unsuccessfully trying to get the core dump.

I hope we're close, and sorry for dragging my feet.

avramd commented:
> Oh you got an integration test done! Awesome! I didn't think that was possible.

It doesn't pass yet, although it is not supposed to. The goal was for it to fail without my patch. It does fail, although I don't know if it's for the right reason yet.

> I'm new to `systemd` PRs, so I didn't know if rebasing was going to make things more difficult. I did it once though, and nobody complained.

Well, @evverx asked you to do a rebase a while ago, so I figured as long as tests pass and the new diff looks the same as the old, it must be legit. I'm new here too.

> 
> I'll run a verification hopefully in the next 24 hours and merge things in. When I tried the patch you suggested last week I was getting crashes... then I lost some time unsuccessfully trying to get the core dump.

I don't know what to say about the crashes, I didn't experience that. However I did this on a different base than your branch in the first place. 

Please don't merge yet, my branch is still a draft. I want to make sure this i-test is fully legit. I'm also adding some minor logging improvements.

> I hope we're close, and sorry for dragging my feet.

It's ok, I hope you don't mind me doing the fork so I could be more productive. This is actually for my day-job, so I'm allowed to use work hours on it.

avramd commented:
@colin-foster-in-advantage, @evverx I've updated the implementation and tests.

_I'm getting build failures from github though, it looks like a fuzz test is failing now. Can either of you offer a tip on how to troubleshoot this? I can't quite make sense of the error message_

Change details:

- We now have a bootp integration test that is passing. Open to suggestions, please at least 👀 my assertions 
  - It was originally failing due to the bug I reported a while back (treating siaddr as "next server" doesn't work for BOOTP)
  - I've fixed that bug
- I reverted one of Colin's changes that was causing false PASSes in his bootp unit tests
  - This was the deliberate skipping of the check for server_address presence when `BOOTP=yes`
  - We now get a good fail in the unit test of the `bootpd` reply packet I submitted. You can verify this in the commit prior to the one containing the bug fix.
- I've temporarily disabled the bootp unit test with Colin's `BOOTREPLY` packet, b/c it is a false negative. 
  - It appears his packet is missing `siaddr`; if I'm correct about that, then it _should_ fail
  - I couldn't figure out how to decode the packets in the unit test, they doen't appear to be in pcap format.
  - I could use a tip or two on how to assert that this test SHOULD fail
- I've added/improved logging
  - We now log when we're interpreting a packet as a BOOTREQUEST, and when sending a BOOTREPLY
  - the `lacks address, server address or lease lifetime...` error log now includes the actual values for those vars

avramd commented:
@colin-foster-in-advantage, @evverx I've updated my branch to reflect the BOOTP reply packet described in the RFE, and to be based on the latest rebase from Colin's fork.

Note that Colin's unit test does not accurately reflect that packet; the test reply packet does not contain DHCP option 54 (Server Identifier). In Colin's branch, it is a false PASS in the unit test b/c his code removes the check for `server_identifier`. Once that error check is re-introduced, that test fails (even without my fix to the BOOTP bug).

I updated that test case to have DHCP Server ID (last commit). I also rewrote my BOOTP fix to use DHCP Option 54 if `siaddr` is missing. I've restored this test case because it does pass now.

Personally I still suspect that is not a valid scenario, and represents a configuration error on a server. I think it would be better to not support it, but I don't mind - I'd rather just get this finished. Either way, we need Colin's input.

colin-foster-in-advantage commented:
> I updated that test case to have DHCP Server ID (last commit). I also rewrote my BOOTP fix to use DHCP Option 54 if `siaddr` is missing. I've restored this test case because it does pass now.
> 
> Personally I still suspect that is not a valid scenario, and represents a configuration error on a server. I think it would be better to not support it, but I don't mind - I'd rather just get this finished. Either way, we need Colin's input.

The only reason I could think that it should be supported is because it works with vanilla `bootpc` which is what I'm replacing.

I agree, let's shoot for the finish line. My https://github.com/colin-foster-in-advantage/systemd/tree/add_bootp_staging branch has `siaddr` or Option 54 logic. I'll squash it into the PR branch and update the documentation (I don't want to force-push-spam the PR).

Then I'll pull in your integration test. @avramd At that point do you think it would be ready?

avramd commented:
Hey @colin-foster-in-advantage, thanks for jumping back in. 

IMHO, `siaddr` is the proper value for `server_address` in a bootp context; opt 54 should only be used to keep the rest of the code happy when `siaddr` is missing.

I'm not sure your bigger picture intentions, so please forgive me if these other comments were already in your plan: Some of the other changes in [my PR](https://github.com/avramd/colin-foster-systemd/pull/1/commits) I believe are worth incorporating:
- Your unit test bootp reply packet doesn't fail when `server_address` is missing
- The existing logging was inappropriately light during error conditions. I think it's only fair to tell admins:
 - When we've activated BOOTP code paths
 - What's going on w/ server_address, since it is conditional and we don't know for sure what is the right choice
 - What the actual problem is in the other error cases of missing address or lease time.

When I first started with all this, it was not possible to tell why it wasn't working from the logs - I had to walk the code. It was reasonable to assume when it said `lease lacks address, server or lease time` that setting `FallbackLeaseLifetimeSec=forever` would be sufficient. Nothing anywhere says that BOOTP is fundamentally not supported, and the `FallbackLeaseLifetimeSec ` docs don't explain that it does not support BOOTP either. Once I added that feature, I got the same error message, and it was basically an incorrect message at that point.
